File: lib/features/faculty/models/faculty.dart
class Faculty {
  final String id;
  final String employeeId;
  final String name;
  final String email;
  final String? phone;
  final String departmentId;
  final String? designation;
  final String? qualification;
  final int? experience;
  final int? joiningDate;
  final String employmentType;
  final String status;
  final String schedulingReadiness;
  final String? skills;
  final String? profilePicture;
  final int createdAt;
  final int updatedAt;

  Faculty({
    required this.id,
    required this.employeeId,
    required this.name,
    required this.email,
    this.phone,
    required this.departmentId,
    this.designation,
    this.qualification,
    this.experience,
    this.joiningDate,
    required this.employmentType,
    required this.status,
    required this.schedulingReadiness,
    this.skills,
    this.profilePicture,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Faculty.fromJson(Map<String, dynamic> json) {
    return Faculty(
      id: json['id'],
      employeeId: json['employeeId'],
      name: json['name'],
      email: json['email'],
      phone: json['phone'],
      departmentId: json['departmentId'],
      designation: json['designation'],
      qualification: json['qualification'],
      experience: json['experience'],
      joiningDate: json['joiningDate'],
      employmentType: json['employmentType'] ?? 'Full Time',
      status: json['status'] ?? 'Active',
      schedulingReadiness: json['schedulingReadiness'] ?? 'Draft',
      skills: json['skills'],
      profilePicture: json['profilePicture'],
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'employeeId': employeeId,
      'name': name,
      'email': email,
      'phone': phone,
      'departmentId': departmentId,
      'designation': designation,
      'qualification': qualification,
      'experience': experience,
      'joiningDate': joiningDate,
      'employmentType': employmentType,
      'status': status,
      'schedulingReadiness': schedulingReadiness,
      'skills': skills,
      'profilePicture': profilePicture,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }
}

class SchedulingProfile {
  final String id;
  final String facultyId;
  final int maxPeriodsPerDay;
  final int maxPeriodsPerWeek;
  final int maxConsecutiveClasses;
  final String? preferredFreeDay;
  final bool avoidFirstHour;
  final bool avoidLastHour;
  final bool canHandleTheory;
  final bool canHandleLabs;
  final bool canHandleTutorials;
  final String? preferredBuildings;
  final int createdAt;
  final int updatedAt;

  SchedulingProfile({
    required this.id,
    required this.facultyId,
    required this.maxPeriodsPerDay,
    required this.maxPeriodsPerWeek,
    required this.maxConsecutiveClasses,
    this.preferredFreeDay,
    required this.avoidFirstHour,
    required this.avoidLastHour,
    required this.canHandleTheory,
    required this.canHandleLabs,
    required this.canHandleTutorials,
    this.preferredBuildings,
    required this.createdAt,
    required this.updatedAt,
  });

  factory SchedulingProfile.fromJson(Map<String, dynamic> json) {
    return SchedulingProfile(
      id: json['id'],
      facultyId: json['facultyId'],
      maxPeriodsPerDay: json['maxPeriodsPerDay'] ?? 4,
      maxPeriodsPerWeek: json['maxPeriodsPerWeek'] ?? 18,
      maxConsecutiveClasses: json['maxConsecutiveClasses'] ?? 2,
      preferredFreeDay: json['preferredFreeDay'],
      avoidFirstHour: json['avoidFirstHour'] ?? false,
      avoidLastHour: json['avoidLastHour'] ?? false,
      canHandleTheory: json['canHandleTheory'] ?? true,
      canHandleLabs: json['canHandleLabs'] ?? true,
      canHandleTutorials: json['canHandleTutorials'] ?? true,
      preferredBuildings: json['preferredBuildings'],
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'facultyId': facultyId,
      'maxPeriodsPerDay': maxPeriodsPerDay,
      'maxPeriodsPerWeek': maxPeriodsPerWeek,
      'maxConsecutiveClasses': maxConsecutiveClasses,
      'preferredFreeDay': preferredFreeDay,
      'avoidFirstHour': avoidFirstHour,
      'avoidLastHour': avoidLastHour,
      'canHandleTheory': canHandleTheory,
      'canHandleLabs': canHandleLabs,
      'canHandleTutorials': canHandleTutorials,
      'preferredBuildings': preferredBuildings,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }
}

class Availability {
  final String id;
  final String facultyId;
  final int dayOfWeek;
  final int period;
  final bool isAvailable;
  final String? reason;
  final int createdAt;
  final int updatedAt;

  Availability({
    required this.id,
    required this.facultyId,
    required this.dayOfWeek,
    required this.period,
    required this.isAvailable,
    this.reason,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Availability.fromJson(Map<String, dynamic> json) {
    return Availability(
      id: json['id'],
      facultyId: json['facultyId'],
      dayOfWeek: json['dayOfWeek'],
      period: json['period'],
      isAvailable: json['isAvailable'] ?? true,
      reason: json['reason'],
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'facultyId': facultyId,
      'dayOfWeek': dayOfWeek,
      'period': period,
      'isAvailable': isAvailable,
      'reason': reason,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }
}

class Leave {
  final String id;
  final String facultyId;
  final String leaveType;
  final int startDate;
  final int endDate;
  final String status;
  final String? reason;
  final int createdAt;
  final int updatedAt;

  Leave({
    required this.id,
    required this.facultyId,
    required this.leaveType,
    required this.startDate,
    required this.endDate,
    required this.status,
    this.reason,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Leave.fromJson(Map<String, dynamic> json) {
    return Leave(
      id: json['id'],
      facultyId: json['facultyId'],
      leaveType: json['leaveType'],
      startDate: json['startDate'],
      endDate: json['endDate'],
      status: json['status'] ?? 'Pending',
      reason: json['reason'],
      createdAt: json['createdAt'],
      updatedAt: json['updatedAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'facultyId': facultyId,
      'leaveType': leaveType,
      'startDate': startDate,
      'endDate': endDate,
      'status': status,
      'reason': reason,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
    };
  }
}

File: backend/models/faculty.py
from sqlalchemy import Column, String, Integer, Boolean, ForeignKey
from core.database import Base

class Faculty(Base):
    __tablename__ = "Faculty"

    id = Column(String, primary_key=True, index=True)
    employeeId = Column("employeeId", String, unique=True, index=True, nullable=False)
    name = Column(String, nullable=False)
    email = Column(String, unique=True, index=True, nullable=False)
    phone = Column(String)
    departmentId = Column("departmentId", String, ForeignKey("Department.id"), nullable=False)
    designation = Column(String)
    qualification = Column(String)
    experience = Column(Integer)  # in years
    joiningDate = Column("joiningDate", Integer)
    employmentType = Column("employmentType", String, default="Full Time")
    status = Column(String, default="Active")
    schedulingReadiness = Column("schedulingReadiness", String, default="Draft")
    skills = Column(String)
    profilePicture = Column("profilePicture", String)
    createdAt = Column("createdAt", Integer)
    updatedAt = Column("updatedAt", Integer)

class CrossDepartmentTeaching(Base):
    __tablename__ = "CrossDepartmentTeaching"

    id = Column(String, primary_key=True, index=True)
    facultyId = Column("facultyId", String, ForeignKey("Faculty.id"), nullable=False)
    departmentId = Column("departmentId", String, ForeignKey("Department.id"), nullable=False)
    createdAt = Column("createdAt", Integer)
    updatedAt = Column("updatedAt", Integer)

class SchedulingProfile(Base):
    __tablename__ = "SchedulingProfile"

    id = Column(String, primary_key=True, index=True)
    facultyId = Column("facultyId", String, ForeignKey("Faculty.id"), unique=True, nullable=False)
    maxPeriodsPerDay = Column("maxPeriodsPerDay", Integer, default=4)
    maxPeriodsPerWeek = Column("maxPeriodsPerWeek", Integer, default=18)
    maxConsecutiveClasses = Column("maxConsecutiveClasses", Integer, default=2)
    preferredFreeDay = Column("preferredFreeDay", String)
    avoidFirstHour = Column("avoidFirstHour", Boolean, default=False)
    avoidLastHour = Column("avoidLastHour", Boolean, default=False)
    canHandleTheory = Column("canHandleTheory", Boolean, default=True)
    canHandleLabs = Column("canHandleLabs", Boolean, default=True)
    canHandleTutorials = Column("canHandleTutorials", Boolean, default=True)
    preferredBuildings = Column("preferredBuildings", String)
    createdAt = Column("createdAt", Integer)
    updatedAt = Column("updatedAt", Integer)

class Availability(Base):
    __tablename__ = "Availability"

    id = Column(String, primary_key=True, index=True)
    facultyId = Column("facultyId", String, ForeignKey("Faculty.id"), nullable=False)
    dayOfWeek = Column("dayOfWeek", Integer, nullable=False)  # 1-7 for Monday-Sunday
    period = Column(Integer, nullable=False)
    isAvailable = Column("isAvailable", Boolean, default=True)
    reason = Column(String)
    createdAt = Column("createdAt", Integer)
    updatedAt = Column("updatedAt", Integer)

class Leave(Base):
    __tablename__ = "Leave"

    id = Column(String, primary_key=True, index=True)
    facultyId = Column("facultyId", String, ForeignKey("Faculty.id"), nullable=False)
    leaveType = Column("leaveType", String, nullable=False)
    startDate = Column("startDate", Integer, nullable=False)
    endDate = Column("endDate", Integer, nullable=False)
    status = Column(String, default="Pending")
    reason = Column(String)
    createdAt = Column("createdAt", Integer)
    updatedAt = Column("updatedAt", Integer)
